### PR TITLE
Check input dimension for contrib.layers.conv2d/conv3d

### LINKED ISSUE #14583: contrib.slim.conv2d doesn't check input dimension
tensorflow version: v1.4.0-rc1-11-g130a514 (installed from pip)

contrib.slim.conv2d doesn't check for input dimensionality. So if you give it a 5D tensor it instead performs a 3D convolution.

Here is a quick way to see this:
```
sess = tf.InteractiveSession()

image_ph = tf.placeholder(tf.float32,
                          shape=(None, None, 10, 10),
                          name='image')
image = tf.expand_dims(image_ph, -1)
conv = tf.contrib.slim.conv2d(image, 3, 3, padding="VALID")

sess.run(tf.global_variables_initializer())
conv_out = conv.eval({image_ph: np.random.random((16, 20, 10, 10))})
```

conv_out will have shape `[16, 18, 8, 8, 3]`, meaning that the time dimension (dim 1) has decreased due to the VALID padding. 

As a side note I think it would be convenient if all conv2d implementations let you specify which dimensions to act on (or acted on the last 3 dims of the input by default). This is useful for example if you have sequential data and want to apply a 2d convolution to each frame independently.

### PR DESCRIPTION
**Pull request by yongtang**
This fix tries to fix the issue raised in #14583 where the input dimension was not checked for contrib.layers.conv2d/conv3d and contrib.slim.conv2d/conv3d.

The issue was that conv2d/conv3d were just aliases of convolution. This fix wrap the conv2d/conv3d with the input dimension check so that incorrect usage will return ValueError.

This fix fixes #14583.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

File: tensorflow/contrib/layers/python/layers/layers.py [hunk 1]
```diff
@@ -1001,70 +1085,111 @@ def convolution(inputs,
     ValueError: If `data_format` is invalid.
     ValueError: Both 'rate' and `stride` are not uniformly 1.
   """
-  if data_format not in [None, 'NWC', 'NCW', 'NHWC', 'NCHW', 'NDHWC', 'NCDHW']:
-    raise ValueError('Invalid data_format: %r' % (data_format,))
+  return _convolution(inputs,
```
Comment (sguada) [inline tensorflow/contrib/layers/python/layers/layers.py]: Why did you create _convolution instead of just using convolution?

**Review by sguada — CHANGES_REQUESTED** [review]

File: tensorflow/contrib/layers/python/layers/layers.py (hunk 1 above)
Comment (yongtang) [reply tensorflow/contrib/layers/python/layers/layers.py]: @sguada Thanks for the review. The `_convolution` has one additional args of `dims=None` (compared to `convolution`) and has the following logic added:
https://github.com/yongtang/tensorflow/blob/bc6128be953fa4163743dad63686d64fdfe720b8/tensorflow/contrib/layers/python/layers/layers.py#L948-L950

In this way, `convolution` will call `_convolution` with `dims=None`, `convolution2d` will call `_convolution` with `dims=2`, and `convolution2d` will call `_convolution` with `dims=3`.

The idea is to leave the api of `convolution/convolution2d/convolution3d` intact, while at the same time only check dims on `convolution2d/convolution3d`

Comment (sguada) [reply tensorflow/contrib/layers/python/layers/layers.py]: Can you add an extra key argument to convolution with default that doesn't change current behavior?

Comment (yongtang) [reply tensorflow/contrib/layers/python/layers/layers.py]: Thanks @sguada. The PR has been updated. Please take a look.

**Review by sguada — CHANGES_REQUESTED** [review]

File: tensorflow/contrib/layers/python/layers/layers.py [hunk 2]
```diff
@@ -993,6 +994,8 @@ def convolution(inputs,
     trainable: If `True` also add variables to the graph collection
       `GraphKeys.TRAINABLE_VARIABLES` (see tf.Variable).
     scope: Optional scope for `variable_scope`.
+    dims: Optional dims (e.g., 2 for Conv 2D) for convolution. Default is None
```
[the remark this answers is not in the document]
Comment (yongtang) [reply tensorflow/contrib/layers/python/layers/layers.py]: Thanks @sguada. The doctoring has been updated.

File: tensorflow/contrib/layers/python/layers/layers.py [hunk 3]
```diff
@@ -932,7 +932,8 @@ def convolution(inputs,
                 variables_collections=None,
                 outputs_collections=None,
                 trainable=True,
-                scope=None):
+                scope=None,
+                dims=None):
```
[the remark this answers is not in the document]
Comment (yongtang) [reply tensorflow/contrib/layers/python/layers/layers.py]: Done.

File: tensorflow/contrib/layers/python/layers/layers.py [hunk 4]
```diff
@@ -1061,10 +1067,91 @@ def convolution(inputs,
       outputs = activation_fn(outputs)
     return utils.collect_named_outputs(outputs_collections, sc.name, outputs)
```
[the remark this answers is not in the document]
Comment (yongtang) [reply tensorflow/contrib/layers/python/layers/layers.py]: @sguada Thanks. Done.

**Comment by yongtang** [comment]
@sguada Thanks for the review. The PR has been updated. Please take a look.

**Review by sguada — APPROVED** [review]

**Pull request merged**